packages/multichain-account-service/src/providers/EvmAccountProvider.ts

```diff
@@ -76,10 +118,64 @@ export class EvmAccountProvider extends BaseBip44AccountProvider {
     return accountsArray;
   }
 
-  async discoverAndCreateAccounts(_: {
+  /**
+   * Discover and create accounts for the EVM provider.
+   *
+   * @param opts - The options for the discovery and creation of accounts.
+   * @param opts.entropySource - The entropy source to use for the discovery and creation of accounts.
+   * @param opts.groupIndex - The index of the group to create the accounts for.
+   * @returns The accounts for the EVM provider.
+   */
+  async discoverAndCreateAccounts(opts: {
     entropySource: EntropySourceId;
     groupIndex: number;
-  }) {
-    return []; // TODO: Implement account discovery.
+  }): Promise<Bip44Account<KeyringAccount>[]> {
+    const provider = this.getEvmProvider();
+    const { entropySource, groupIndex } = opts;
+
+    const [address, didCreate] = await this.#createAccount({
+      entropySource,
+      groupIndex,
+    });
+
+    // We don't want to remove the account if it's the first one.
+    const shouldCleanup = didCreate && groupIndex !== 0;
+    try {
+      const countHex = (await provider.request({
+        method: 'eth_getTransactionCount',
+        params: [address, 'latest'],
+      })) as Hex;
+      const count = parseInt(countHex, 16);
+
+      if (count === 0 && shouldCleanup) {
+        await this.withKeyring<EthKeyring>(
+          { id: entropySource },
+          async ({ keyring }) => {
+            keyring.removeAccount?.(address);
+          },
+        );
+        return [];
+      }
+    } catch {
+      // If the RPC request fails and we just created this account for discovery,
+      // remove it to avoid leaving a dangling account.
+      if (shouldCleanup) {
+        await this.withKeyring<EthKeyring>(
+          { id: entropySource },
+          async ({ keyring }) => {
+            keyring.removeAccount?.(address);
+          },
+        );
+      }
+      return [];
```

Comment:
I wonder if we should re-throw here actually... We already handle errors in the discovery orchestrator. This way we would log the error.

If we return `[]` this will be seen as a "nothing to discover" basically.